src/merkle_tree.rs

```diff
@@ -91,18 +91,20 @@ impl MerkleTree {
             let current_layer = &mut self.layers[curr_idx];
             let previous_layer_elem_count = previous_layer.len();
 
-            if (previous_layer.len() - 1) % 2 == 0 {
-                // Duplicate left node
-                hasher.update(previous_layer[previous_layer_elem_count - 1]);
-                hasher.update(previous_layer[previous_layer_elem_count - 1]);
-                current_layer.push(hasher.finalize_reset().into());
-            } else {
+            // If after adding an element the previous layer has even amount of nodes we
+            // don't need to duplicate but if its odd we have to duplicate the left node
+            if previous_layer.len() % 2 == 0 {
                 hasher.update(previous_layer[previous_layer_elem_count - 2]);
                 hasher.update(previous_layer[previous_layer_elem_count - 1]);
                 let Some(last_element) = current_layer.last_mut() else {
                     panic!();
                 };
```

Comment:
Why can this be None? I would add a comment clarifying because it's not immediately obvious.

Reply:
You can use `.expect("last_element should never be None")` or something like that

Reply:
Done